plugins/search-backend-node/src/collators/NewlineDelimitedJsonCollatorFactory.ts

```diff
@@ -0,0 +1,147 @@
+/*
+ * Copyright 2021 The Backstage Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { UrlReader } from '@backstage/backend-common';
+import { Config } from '@backstage/config';
+import { Permission } from '@backstage/plugin-permission-common';
+import { DocumentCollatorFactory } from '@backstage/plugin-search-common';
+import { parse as parseNdjson } from 'ndjson';
+import { Readable } from 'stream';
+import { Logger } from 'winston';
+
+/**
+ * @beta
+ */
+export type NewlineDelimitedJsonCollatorFactoryOptions = {
+  type: string;
+  searchPattern: string;
+  reader: UrlReader;
+  logger: Logger;
+  visibilityPermission?: Permission;
+};
+
+/**
+ * Factory class producing a collator that can be used to index documents
+ * sourced from the latest newline delimited JSON file matching a given search
+ * pattern. "Latest" is determined by the name of the file (last alphabetically
+ * is considered latest).
+ *
+ * @remarks
+ * The reader provided must implement the `search()` method as well as the
+ * `readUrl` method whose response includes the `stream()` method. Naturally,
+ * the reader must also be configured to understand the given search pattern.
+ *
+ * @example
+ * Here's an example configuration using Google Cloud Storage, which would
+ * return the latest file under the `bucket` GCS bucket with files like
+ * `xyz-2021.ndjson` or `xyz-2022.ndjson`.
+ * ```ts
+ * indexBuilder.addCollator({
+ *   schedule,
+ *   factory: NewlineDelimitedJsonCollatorFactory.fromConfig(env.config, {
+ *     type: 'techdocs',
+ *     searchPattern: 'https://storage.cloud.google.com/bucket/xyz-*',
```

Comment:
Do you see any use cases where you might need to pass more than one `searchPattern`?
What about being an array and also adding a property to exclude files that accept a regex or a predicate function?

Reply:
> _Do you see any use cases where you might need to pass more than one searchPattern?_

Good forwarding-thinking!  Hopefully just the pattern as a string will be sufficient.  Globs can be very flexible, and we're somewhat limited by the `UrlReader.search()` method.

Probably if you're using this collator, you're regularly generating `.ndjson` files, and it might just be easier/better to change where/how those files are stored, then it would be to try and really tweak a glob or regex pattern to match just the right files.

> _...or a predicate function?_

This could come in handy to provide custom logic for determining what's "latest?" (vs. doing alphanumeric order), eh?  But maybe we can leave it as a future improvement for someone with the need.